### DISCUSSION TITLE
Geo location information not being saved, sometimes

### DISCUSSION THREAD

**Discussion by keif888** [photoprism/photoprism#4502] [Q&A] [answered]
**Should I raise this as a bug?**

### Issue
When using the Chrome Extension "Geolocation Plugin for PhotoPrism" sometimes the location information is not saved to one or more of the photos selected for a specific location.
I have checked if the extension is the issue, but it is sending valid and correct api calls to PhotoPrism to update the location.
Sometimes the returned json message from PhotoPrism does not have the updated longitude and latitude information, and the sidecar files and the database are not updated.

### Diagnosing the issue
When attempting to replicate this against a newly installed PhotoPrism instance with debug logging turned on, on the 2 occasions that I have been able to replicate it I noticed the following:

**Successful update**
```
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="place: dk:5UOSZKIepmrZ not found"
photoprism-1  | time="2024-09-04T12:25:29Z" level=info msg="cell: added place dk:5UOSZKIepmrZ [1.400746011s]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="cell: added s2:465253379adc [1.403718376s]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="photo: changing location of 2024/09/20240904_120425_C446A87E from s2:464ff36f20ec to s2:465253379adc"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="photo: 2024/09/20240904_120425_C446A87E has new title 'København S / Denmark / 2024' [1.116321ms]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="search: updated index by photo id [1.162236ms]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="index: updating counts"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="counts: updated 1 place [949.567µs]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="counts: updated 0 subjects [628.424µs]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="counts: updated 0 labels [477.931µs]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="photos: found 1 result for uid:psjaew32utj74kee dist:2.000000 merged:true [1.05134ms]"
photoprism-1  | time="2024-09-04T12:25:29Z" level=info msg="changes successfully saved"
photoprism-1  | time="2024-09-04T12:25:29Z" level=info msg="photo: updated sidecar file 2024/09/20240904_120425_C446A87E.yml"
photoprism-1  | time="2024-09-04T12:25:29Z" level=debug msg="server: PUT /api/v1/photos/psjaew32utj74kee (200) [1.472165736s]"
```

**Failed Update**
Note the warning - Deadlock message instead of the counts: updated 0 labels above.
```
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="place: gb:u7WNUrVi23K5 not found"
photoprism-1  | time="2024-09-04T12:28:12Z" level=info msg="cell: added place gb:u7WNUrVi23K5 [1.290580231s]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="cell: added s2:487604ce4af4 [1.292516235s]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="photo: changing location of 2024/09/20240904_120425_406A1B87 from s2:465253379adc to s2:487604ce4af4"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="photo: 2024/09/20240904_120425_406A1B87 title based on label Memorial"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="photo: 2024/09/20240904_120425_406A1B87 has new title 'Memorial / Westminster / 2024' [1.253884ms]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="search: updated index by photo id [1.135424ms]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="index: updating counts"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="counts: updated 1 place [1.633586ms]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="counts: updated 0 subjects [644.176µs]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=warning msg="index: Error 1213 (40001): Deadlock found when trying to get lock; try restarting transaction while updating label counts (update counts)"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="photos: found 1 result for uid:psjaew5oxefktsbe dist:2.000000 merged:true [1.219616ms]"
photoprism-1  | time="2024-09-04T12:28:12Z" level=info msg="changes successfully saved"
photoprism-1  | time="2024-09-04T12:28:12Z" level=info msg="photo: updated sidecar file 2024/09/20240904_120425_406A1B87.yml"
photoprism-1  | time="2024-09-04T12:28:12Z" level=debug msg="server: PUT /api/v1/photos/psjaew5oxefktsbe (200) [1.40565443s]"
```

This looks like a bug to me, where the deadlock is preventing the database and sidecar file from being updated, and then the json response is returning the original data (because the updates were not done).
To me this should not be returning a http 200, but an appropriate error maybe 500 Server Error?

This bug is hard to replicate, as you have to be "unlucky" to get a deadlock when applying an update.

(Please note that I have a pull request to add retries into the Geolocation Plugin for PhotoPrism to attempt to work around this issue https://github.com/andyvalerio/photoprism-geolocation/pull/6
This checks if the location information in the returned json is updated, and retries up to 5 times if not.)

**Comment by lastzero** [comment]
What database type and version are you using? Does it work if you use the latest MariaDB release?

**Reply by keif888** [reply]
Brand new install using https://dl.photoprism.app/docker/docker-compose.yml
mariadb-1     | 2024-09-04 13:33:48+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.5.2+maria~ubu2404 started.
p
PhotoPrism Build [240711-2197af848](https://docs.photoprism.app/release-notes/)

For my testing I am using VirtualBox hosted Debian 12.7 fresh installed today, 4 core CPU, 8Gb Ram, SSD disk.

**Reply by lastzero** [reply]
PhotoPrism requests the location information from a service we provide (reverse geocoding). This request may fail in some cases, e.g. because of a server restart or because your internet connection was unstable. You may want to check that if the plugin and database are not causing the problem. Just to be sure: You are not accessing the SSD over a network connection?

**Reply by keif888** [reply]
I have found a way to replicate this issue consistently, although it's not using deadlocks but waitlocks.  This replication does not require the Chrome plugin, any browser will do.

The underlying issue, in my view, is that photoprism is not returning errors just aborting updates when there is a database level issue during the api call, and it is logging that the updates were successful even though they are not.
I don't know the GO language so I haven't tried to determine where in the code this is going wrong.  But I suspect the issue is in [here](https://github.com/photoprism/photoprism/blob/1d12bb62f07eb895a1b6dd4dc07803909b568723/internal/api/photos.go#L101) somewhere.

### Replicate Issue
This assumes that you can select/update data from the mariadb database.  I used the photoprism user and password from my docker-compose.yml file, and docker inspect against the mariadb container to find the ip address to connect to.

1. choose a photo that you will update, and note the labels that are applied to it
2. choose a label that you will use to simulate the issue, and find it's id.  Replace < value > with the label name you chose.
```
select id, label_name, photo_count from photoprism.labels where label_name = '<value>';

```
3. edit the photo
4. change the longitude and latitude, but do NOT press APPLY or DONE.
5. issue the following sql commands replacing the < id > with the number of the label that you chose above:
```
set autocommit = 0;
begin work;
update photoprism.labels
set photo_count = 2
where id = <id>;
```
6. press DONE in the photoprism app
7. wait for the lock wait timeout exceeded that will occur (this took 4 minutes in my testing)
8. the photoprism app should now have returned to the library, without any errors shown
9. open the selected photo, and note that the longitude and latitude have not been saved
10. rollback the database update
```
rollback work;
```

The instructions above resulted in the following logs.
```
photoprism-1  | time="2024-09-05T05:59:49Z" level=debug msg="server: GET /api/v1/photos/psjagnyruxdjht49 (200) [5.706074ms]"
photoprism-1  | time="2024-09-05T06:05:22Z" level=debug msg="place: se:e5AOkRzxrfVf not found"
photoprism-1  | time="2024-09-05T06:05:22Z" level=info msg="cell: added place se:e5AOkRzxrfVf [1.136678896s]"
photoprism-1  | time="2024-09-05T06:05:22Z" level=debug msg="cell: added s2:464ff390c904 [1.138055982s]"
photoprism-1  | time="2024-09-05T06:05:22Z" level=debug msg="photo: changing location of 2024/09/20240904_124225_40F1A895 from zz to s2:464ff390c904"
photoprism-1  | time="2024-09-05T06:05:22Z" level=debug msg="photo: 2024/09/20240904_124225_40F1A895 title based on label University"
photoprism-1  | time="2024-09-05T06:05:22Z" level=debug msg="photo: 2024/09/20240904_124225_40F1A895 has new title 'University / Göteborg / 2024' [763.25µs]"
photoprism-1  | time="2024-09-05T06:07:09Z" level=debug msg="index: updating metadata"
photoprism-1  | time="2024-09-05T06:07:22Z" level=debug msg="search: updated index by photo id [915.031µs]"
photoprism-1  | time="2024-09-05T06:07:22Z" level=debug msg="index: updating counts"
photoprism-1  | time="2024-09-05T06:07:22Z" level=debug msg="counts: updated 1 place [1.38752ms]"
photoprism-1  | time="2024-09-05T06:07:22Z" level=debug msg="counts: updated 0 subjects [461.779µs]"
photoprism-1  | time="2024-09-05T06:09:22Z" level=warning msg="index: Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction while updating label counts (update counts)"
photoprism-1  | time="2024-09-05T06:09:22Z" level=debug msg="photos: found 1 result for uid:psjagnyruxdjht49 dist:2.000000 merged:true [1.445978ms]"
photoprism-1  | time="2024-09-05T06:09:22Z" level=info msg="changes successfully saved"
photoprism-1  | time="2024-09-05T06:09:22Z" level=info msg="photo: updated sidecar file 2024/09/20240904_124225_40F1A895.yml"
photoprism-1  | time="2024-09-05T06:09:22Z" level=debug msg="server: PUT /api/v1/photos/psjagnyruxdjht49 (200) [4m1.302819537s]"
```

**Reply by lastzero** [reply]
I happened to check the MariaDB Release Notes yesterday and it seems that some critical bugs related to locking have been fixed:
- https://mariadb.com/kb/en/mariadb-11-5-2-release-notes/#innodb

Maybe the issue is solved when you upgrade to MariaDB 11.5.2? A new Docker image for it has just been released:
- https://hub.docker.com/_/mariadb/tags?page=&page_size=&ordering=&name=11.5.2

**Reply by lastzero** [reply]
My apologies, I noticed that you already seem to be using 11.5.2. In that case, I hope it's a problem with our code and not another MariaDB issue.... I should know by next week as I've just upgraded our servers.

**Reply by lastzero** [reply]
If you have not increased the default buffer pool size, I recommend you try this to see if it improves the situation:
- https://docs.photoprism.app/getting-started/troubleshooting/performance/#mariadb

**Reply by keif888** [reply]
I'm from a database administration background, and MariaDB is working as I would expect it to, given what PhotoPrism and I are doing.

I'm trying to work through what the api is doing with the update.  My not knowing GO as a language is not helping.
I did a database trace to see what is being sent to the database.

This set of database statements is silently failing (no log messages are being raised in the GO code), and it is where the location update is being lost because there is a "lock/deadlock" on the labels table.
I think this is being driven from here:
https://github.com/photoprism/photoprism/blob/1d12bb62f07eb895a1b6dd4dc07803909b568723/internal/entity/photo.go#L200
but I'm not sure and haven't managed to get a developer copy of PhotoPrism going so I can trace it.
I can't find any rollback statements in the code, so I'm assuming that GORM is taking care of that under the covers.

```
2024-09-05 07:23:18.912152	photoprism[photoprism] @  [172.19.0.3]	23	1	Query	START TRANSACTION
2024-09-05 07:23:18.912646	photoprism[photoprism] @  [172.19.0.3]	23	1	Prepare	UPDATE `photos` SET `camera_id` = ?, `camera_serial` = ?, `camera_src` = ?, `cell_accuracy` = ?, `cell_id` = ?, `checked_at` = ?, `created_by` = ?, `deleted_at` = ?, `description_src` = ?, `edited_at` = ?, `estimated_at` = ?, `lens_id` = ?, `original_name` = ?, `photo_altitude` = ?, `photo_color` = ?, `photo_country` = ?, `photo_day` = ?, `photo_description` = ?, `photo_duration` = ?, `photo_exposure` = ?, `photo_f_number` = ?, `photo_faces` = ?, `photo_favorite` = ?, `photo_focal_length` = ?, `photo_iso` = ?, `photo_lat` = ?, `photo_lng` = ?, `photo_month` = ?, `photo_name` = ?, `photo_panorama` = ?, `photo_path` = ?, `photo_private` = ?, `photo_quality` = ?, `photo_resolution` = ?, `photo_scan` = ?, `photo_stack` = ?, `photo_title` = ?, `photo_type` = ?, `photo_year` = ?, `place_id` = ?, `place_src` = ?, `published_at` = ?, `taken_at` = ?, `taken_at_local` = ?, `taken_src` = ?, `time_zone` = ?, `title_src` = ?, `type_src` = ?, `updated_at` = ?, `uuid` = ?  WHERE `photos`.`id` = ?
2024-09-05 07:23:18.912815	photoprism[photoprism] @  [172.19.0.3]	23	1	Execute	UPDATE `photos` SET `camera_id` = 1, `camera_serial` = '', `camera_src` = '', `cell_accuracy` = 0, `cell_id` = 'zz', `checked_at` = '2024-09-04 13:08:34', `created_by` = 'usjaefwhstvplw6i', `deleted_at` = NULL, `description_src` = '', `edited_at` = '2024-09-05 07:23:18', `estimated_at` = NULL, `lens_id` = 1, `original_name` = 'vitaliy-shevchenko-gR1V3b48Z8s-unsplash', `photo_altitude` = 0, `photo_color` = 6, `photo_country` = 'au', `photo_day` = -1, `photo_description` = '', `photo_duration` = 0, `photo_exposure` = '', `photo_f_number` = 0, `photo_faces` = 0, `photo_favorite` = 0, `photo_focal_length` = 0, `photo_iso` = 0, `photo_lat` = 0, `photo_lng` = 0, `photo_month` = -1, `photo_name` = '20240904_124225_40F1A895', `photo_panorama` = 0, `photo_path` = '2024/09', `photo_private` = 0, `photo_quality` = 3, `photo_resolution` = 24, `photo_scan` = 0, `photo_stack` = 0, `photo_title` = 'Vitaliy Shevchenko Gr S Unsplash', `photo_type` = 'image', `photo_year` = -1, `place_id` = 'zz', `place_src` = 'manual', `published_at` = NULL, `taken_at` = '2024-09-04 06:42:25', `taken_at_local` = '2024-09-04 14:42:25', `taken_src` = '', `time_zone` = 'Australia/Perth', `title_src` = '', `type_src` = '', `updated_at` = '2024-09-05 07:23:18.912304691', `uuid` = ''  WHERE `photos`.`id` = 31
2024-09-05 07:23:18.913125	photoprism[photoprism] @  [172.19.0.3]	23	1	Close stmt	
2024-09-05 07:23:18.913353	photoprism[photoprism] @  [172.19.0.3]	23	1	Prepare	UPDATE `labels` SET `label_uid` = ?, `label_slug` = ?, `custom_slug` = ?, `label_name` = ?, `label_priority` = ?, `label_favorite` = ?, `label_description` = ?, `label_notes` = ?, `photo_count` = ?, `thumb` = ?, `thumb_src` = ?, `created_at` = ?, `updated_at` = ?, `published_at` = ?, `deleted_at` = ?  WHERE `labels`.`deleted_at` IS NULL AND `labels`.`id` = ?
2024-09-05 07:23:18.913506	photoprism[photoprism] @  [172.19.0.3]	23	1	Execute	UPDATE `labels` SET `label_uid` = 'lsjagiw6wiqilm9b', `label_slug` = 'university', `custom_slug` = 'university', `label_name` = 'University', `label_priority` = -1, `label_favorite` = 0, `label_description` = '', `label_notes` = '', `photo_count` = 31, `thumb` = '85b82010f0981add5bd94f15cbeccfdf553af481', `thumb_src` = '', `created_at` = '2024-09-04 12:40:08', `updated_at` = '2024-09-05 07:23:18.913195274', `published_at` = NULL, `deleted_at` = NULL  WHERE `labels`.`deleted_at` IS NULL AND `labels`.`id` = 9
2024-09-05 07:25:18.913814	photoprism[photoprism] @  [172.19.0.3]	23	1	Close stmt	
2024-09-05 07:25:18.914024	photoprism[photoprism] @  [172.19.0.3]	23	1	Query	ROLLBACK
```
You can see that there is a 2 minute delay between the labels update and the rollback, which is the timeout happening.

After that there is another attempt to update the labels, which generates a log message.  This is from this statement in the SavePhotoForm.  Although this fails, it just logs the failure.  But by this point the update to the photos table has already failed.
https://github.com/photoprism/photoprism/blob/1d12bb62f07eb895a1b6dd4dc07803909b568723/internal/entity/photo.go#L205

The following is a combination of the database log and the photoprism log.
```
2024-09-05 07:25:18.919537	photoprism[photoprism] @  [172.19.0.3]	23	1	Query	UPDATE labels LEFT JOIN (
		SELECT p2.label_id, COUNT(DISTINCT photo_id) AS label_photos FROM (
			SELECT pl.label_id as label_id, p.id AS photo_id FROM photos p
				JOIN photos_labels pl ON pl.photo_id = p.id AND pl.uncertainty < 100
			WHERE p.photo_quality > -1 AND p.photo_private = 0 AND p.deleted_at IS NULL
			UNION
			SELECT c.category_id as label_id, p.id AS photo_id FROM photos p
				JOIN photos_labels pl ON pl.photo_id = p.id AND pl.uncertainty < 100
				JOIN categories c ON c.label_id = pl.label_id
			WHERE p.photo_quality > -1 AND p.photo_private = 0 AND p.deleted_at IS NULL
			) p2 GROUP BY p2.label_id
		) b ON b.label_id = labels.id
		SET photo_count = CASE WHEN b.label_photos IS NULL THEN 0 ELSE b.label_photos END
photoprism-1  | time="2024-09-05T07:27:18Z" level=warning msg="index: Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction while updating label counts (update counts)"
```

**Reply by keif888** [reply]
I have found the underlying bug.
It is in entity_update.go line 32.
https://github.com/photoprism/photoprism/blob/98b0dc5e07ac80fe83b439478fe767f4760bc01a/internal/entity/entity_update.go#L32

```
	// Update values.
	result := db.Model(m).Updates(values)

	// Successful?
	if result.Error != nil {
		return err  // <<--------  This is wrong!!!
	} else if result.RowsAffected > 1 {
		log.Debugf("entity: updated statement affected more than one record - you may have found a bug")
		return nil
```
It should be returning the result.Error value, as err is not populated by the preceding database call, result is instead.

```
	// Update values.
	result := db.Model(m).Updates(values)

	// Successful?
	if result.Error != nil {
		return result.Error
	} else if result.RowsAffected > 1 {
		log.Debugf("entity: updated statement affected more than one record - you may have found a bug")
		return nil
```

Making this change will propagate the error back up the call stack to internal/api/photos.go, which will issue a http 500 response back to the caller of the api.
There are also be a couple of other attempts in the code to retry the call, so in a deadlock situation it may (hopefully) self resolve.

**Reply by keif888** [reply, marked as answer]
I have raised issue #4504 about this.
